Sources/CodableKit/Reflection/HiLoDecoder.swift

```diff
@@ -112,62 +117,68 @@ struct HiLoDecoder: Decoder {
         }
         let ctx: Context
         let codingPath: [CodingKey]
-        
-        init(_ ctx: Context, codingPath: [CodingKey]) {
+        let keyPath: KeyPath<Root, Value>?
+
+        init(_ ctx: Context, codingPath: [CodingKey], keyPath: KeyPath<Root, Value>?) {
             self.ctx = ctx
             self.codingPath = codingPath
             self.isAtEnd = false
             self.currentIndex = 0
+            self.keyPath = keyPath
         }
-        
+
         mutating func decodeNil() throws -> Bool {
             ctx.nextIsOptional = true
             return false
         }
-        
+
         mutating func decode<T>(_ type: T.Type) throws -> T where T : Decodable {
             isAtEnd = true
-            return try SingleValueDecoder(ctx, codingPath: codingPath + [key]).decode(T.self)
+            return try SingleValueDecoder(ctx, codingPath: codingPath + [key], keyPath: keyPath).decode(T.self)
         }
-        
+
         mutating func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type) throws -> KeyedDecodingContainer<NestedKey> where NestedKey : CodingKey {
-            return .init(KeyedDecoder<NestedKey>(ctx, codingPath: codingPath + [key]))
+            return .init(KeyedDecoder<NestedKey>(ctx, codingPath: codingPath + [key], keyPath: keyPath))
         }
-        
+
         mutating func nestedUnkeyedContainer() throws -> UnkeyedDecodingContainer {
-            return UnkeyedDecoder(ctx, codingPath: codingPath + [key])
+            return UnkeyedDecoder(ctx, codingPath: codingPath + [key], keyPath: keyPath)
         }
-        
+
         mutating func superDecoder() throws -> Decoder {
-            return HiLoDecoder(ctx, codingPath: codingPath + [key])
+            return HiLoDecoder(ctx, codingPath: codingPath + [key], keyPath: keyPath)
         }
     }
-    
+
     private struct SingleValueDecoder: SingleValueDecodingContainer {
         let ctx: Context
         let codingPath: [CodingKey]
-        
-        init(_ ctx: Context, codingPath: [CodingKey]) {
+        let keyPath: KeyPath<Root, Value>?
+
+        init(_ ctx: Context, codingPath: [CodingKey], keyPath: KeyPath<Root, Value>?) {
             self.ctx = ctx
             self.codingPath = codingPath
+            self.keyPath = keyPath
         }
-        
+
         func decodeNil() -> Bool {
             ctx.nextIsOptional = true
             return false
         }
-        
+
         func decode<T>(_ type: T.Type) throws -> T where T : Decodable {
             ctx.add(T.self, at: codingPath)
-            if let custom = T.self as? AnyReflectionDecodable.Type {
+
+            if let custom = T.self as? AnyReflectionDecodable.Type,
+                keyPath?.valueType == T.self || custom.isBaseType {
```

Comment:
You should be able to do `Root.self == T.self` here without needing to pass around the `keyPath`.